### ISSUE TITLE
Can termux access phone's bluetooth adapter?

### ISSUE THREAD

**Issue by x92rwzb** [termux/termux-app#235]
I recently started using termux and I can't seem to find a way to access my phone's bluetooth adapter. Is it even possible?

**Comment by fornwall** [comment]
A command could perhaps be added to the Termux:API add-on app.

Do you have an idea of how it would look like? Perhaps something like

    termux-bluetooth-scan

to find out addresses, and

    termux-bluetooth-connect <address>

to connect to a device, communicating over stdout&stdin?

**Comment by x92rwzb** [comment]
I don't think it would it be possible to use third party tools with that approach though.

**Comment by martmists** [comment]
Having a library that supports Bluetooth headers would be a great idea in my opinion, as e.g. `pybluez` needs `bluetooth/bluetooth.h`, and I'm using quite a lot of projects that depend on pybluez.

**Comment by wayofthefuture** [comment]
Is it a complex problem to add the ability to stream data from a Bluetooth device? Does Termux not have access to this device after installing? Just trying to understand the problem and possible solutions. Thanks.

**Comment by wayofthefuture** [comment]
I am able to see the serial port devices under the root drive "/dev/tty...", does this mean I can read data from these ports from the Termux shell or running app? We have NodeJS installed on Termux. Thanks.

**Comment by tomtom** [comment]
Would it be possible to have a command to switch bluetooth on and off again? I personally don't need to be able to configure bluetooth from the command line.

**Comment by Grimler91** [comment]
@tomtom what's the purpose of switching it on and off?

**Comment by tomtom** [comment]
Basic use case: have bluetooth turned off by default but turn it on when starting termux -- and maybe turn it off again when quitting termux.

**Comment by Grimler91** [comment]
@tomtom that makes sense. 

I might be possible already today to turn on/off bluetooth using am, but would require root unless some permissions are added to the app I think. See https://stackoverflow.com/questions/37259260/android-enable-disable-bluetooth-via-command-line

**Comment by showking** [comment]
> A command could perhaps be added to the Termux:API add-on app.
> 
> Do you have an idea of how it would look like? Perhaps something like
> 
> ```
> termux-bluetooth-scan
> ```
> to find out addresses, and
> 
> ```
> termux-bluetooth-connect <address>
> ```
> to connect to a device, communicating over stdout&stdin?

I like this idea. What is needed to implement?

**Comment by blackprojects** [comment]
Maybe anybody can take code from https://github.com/StevenSalazarM/Termux-api-bluetooth and integrate? I try this package, but always has error - "unused DT entry: type 0xXX arg"

**Comment by xeffyr** [comment]
> Maybe anybody can take code

No, we can't. At least because it will not work without corresponding API implementation on [Termux:API](https://github.com/termux/termux-api) application's side.

**Comment by blackprojects** [comment]
@xeffyr do you think about implement?

**Comment by martmists** [comment]
This implementation (in my opinion) lacks the `bluetooth.h` header one would expect from bluetooth support.

**Comment by xeffyr** [comment]
@martmists This is not "implementation" at all. Just a clone of https://github.com/termux/termux-api-package with few more scripts.

**Comment by xeffyr** [comment]
For those who seeking for implementation, it is in https://github.com/termux/termux-app/pull/517.

**Comment by RodrigoTomeES** [comment]
Hi,

There is any update about this feature?

Thanks!

**Comment by Munchax10** [comment]
I made [a pull request](https://github.com/termux/termux-api/pull/686) for this, please merge it.
